crates/runtime-table-partition/src/provider/pruning.rs

```diff
@@ -15,96 +15,535 @@ limitations under the License.
+                            partition_by,
+                            partition_value,
+                            &lit,
+                        )?;
+                        any_matches |= is_match;
+                    }
+                    if !any_matches {
+                        return Ok(true);
+                    }
+                } else if let Some((col_name, literals)) = collect_and_inequalities(filter) {
+                    for lit in literals {
+                        let is_match = filter_or_udf_value_matches(
+                            &Expr::Column(col_name.clone().into()),
+                            partition_by,
+                            partition_value,
+                            &lit,
+                        )?;
+                        if is_match {
+                            return Ok(true); // Prune if match for NOT IN-like condition
+                        }
+                    }
+                }
+            }
+            Expr::InList(InList {
+                expr,
+                list,
+                negated,
+            }) => {
+                if let Expr::Column(_) = expr.as_ref() {
+                    let mut any_matches = false;
+                    for lit in list {
+                        if let Expr::Literal(lit_val) = lit {
+                            let is_match = filter_or_udf_value_matches(
+                                expr,
+                                partition_by,
+                                partition_value,
+                                lit_val,
+                            )?;
+                            if is_match && *negated {
+                                return Ok(true); // prune if match in NOT IN
+                            }
+                            any_matches |= is_match;
+                        }
+                    }
+                    if !any_matches && !negated {
+                        return Ok(true);
+                    }
                 }
             }
+            _ => {}
         }
     }
-    false
+
+    Ok(false)
+}
+
+fn filter_or_udf_value_matches(
+    column: &Expr,
+    partition_by: &Expr,
+    partition_value: &ScalarValue,
+    filter_value: &ScalarValue,
+) -> Result<bool, DataFusionError> {
+    let Expr::Column(col) = column else {
+        return Err(DataFusionError::Plan("Expected column expression".into()));
+    };
+
+    // Replace column reference with filter value in partition_by expression
+    let transformed_expr = partition_by
+        .clone()
+        .transform(|e| {
+            Ok(match e {
+                Expr::Column(expr_col) if expr_col == *col => {
+                    Transformed::yes(Expr::Literal(filter_value.clone()))
+                }
+                _ => Transformed::no(e),
+            })
+        })
+        .map_err(|e| DataFusionError::Plan(format!("Failed to transform expression: {e}")))?
+        .data;
+
+    let result = match transformed_expr {
```

Comment:
This still isn't executing the expression - this won't handle more advanced partition_by expressions like `CASE ... WHEN... THEN...`. I'll merge this for release, but I would want to see a general expression evaluator being used here.